### ISSUE TITLE
GoogleMiddleware: Correlation failed

### ISSUE THREAD

**Issue by hempels** [aspnet/Security#1130]
We have significant numbers of users are unable to login using Google auth because whenever they try, they fail with messages logged like:

Microsoft.AspNetCore.Authentication.Google.GoogleMiddleware[15]
      '.AspNetCore.Correlation.Google.dP7ln8REPkhVI0h1NeWykBU3ev5Msy0hIuAxAyfgP3I' cookie not found.
Microsoft.AspNetCore.Authentication.Google.GoogleMiddleware[4]
      Error from RemoteAuthentication: Correlation failed..

I dug into the code and found RemoteAuthenticationTimeout, so we tried setting that value to 24 hours to ensure no one would hit this due to a cookie expiration. Even with that setting, there was no decrease in the number of users experiencing the issue. It only seems to be happening when they use Chrome, so we suspect it might be an extension or antivirus causing an issue.

The problem is, we can't disable the check and we can't tell users not to run extensions. I have been unable to come up with any scenarios where not doing that check would introduce a security risk in our implementation, so I would like to simply disable it. However, there is no apparent option to do so.

Therefore, this is a question and depending on the answer, a bug/feature request. Is it possible currently to disable the remote authentication correlation cookie validation? If not, please provide an option for that. This isn't a small number of users affected, it's currently preventing somewhere around 20% of our user base from using Chrome to run our application.

We are running v1.1.0

**Comment by hempels** [comment]
We finally found a reproducible scenario.

In our implementation, we pass ?prompt=select_account which forces Google to take the user to their Account Chooser page on every authentication. While there, if the user clicks on their account multiple times (not a double-click, but multiple single clicks) then multiple requests are sent through RemoteAuthentication. Because the first such request deletes the Correlation Cookie, all subsequent requests fail.

We obviously can't control users' behavior, nor can we change Google's Account Chooser page to prevent multiple clicks (although we will request it.) So that leaves modifying the ValidateCorrelationId behavior as the only option we have to fix this.

I'm working on a patch that we can run, but the simple solution is that there's no reason to delete those correlation cookies after they are validated. They have an explicit expiration set, so they will delete themselves.

**Comment by Tratcher** [comment]
When the user clicks twice does it cause the first request to be canceled (or the response ignored) in favor of the second one?

**Comment by hempels** [comment]
The first response is ignored, the second one goes through and fails.

**Comment by Tratcher** [comment]
That's problematic, it makes it look like a replay attack. Let's pursue this with Google first. Please link to your Google bug report so we can follow along.

**Comment by hempels** [comment]
Replay attack concern understood, but difficult to pull off. There's already a potential attack during the time the cookie is set and waiting for the callback redirect. Shortening the default cookie expiration time from 15 minutes to something like 2 minutes would help with both cases. Also you could add a flag (AllowReplay) to RemoteAuthenticationOptions which would make it the consumer's choice whether to open up that attack vector and not the framework's responsibility.

Given that I don't expect Google to react quickly to an issue report, I'm focusing on fixing this for our users before I figure out how to file it.

**Comment by Eilon** [comment]
Unfortunately given how the protocol works there isn't a way for this to be fixed within ASP.NET Core. Once the cookie is used up, both the cookie and all the data in it are "used up" and cannot be replayed. We considered ways of extending the duration of the cookie but that wouldn't work due to single-use codes.

But, you can handle the AuthenticationFailed event to detect this case and redirect the user back to the login page and they can take another crack at logging in with Google's provider.

We do recommend filing an issue with Google to have them consider changing the approval UI to be less prone to double-clicking.

**Comment by hempels** [comment]
There's the Microsoft we love: "Not our problem."

There's no reason you can't add an option to allow replays. I understand that arguably violates the OAuth protocol, but allowing consumers to opt in to that to work around a problematic provider seems very reasonable.

Meanwhile I'll be forced to continue to run my own build of the Authentication libraries to do just that. And those not able to figure out how to do that will be stuck with users unable to log in.

**Comment by Eilon** [comment]
@hempels allowing replays of the tokens is potentially a serious security vulnerability. There are good reasons why the protocols don't allow this.

We certainly realize this is unfortunate, but risking compromising the security of the system is not ideal.

Our recommendation is two things:
1. The app can handle the specific error case where the token has been used up and ask the user to sign in again (with Google).
2. Have Google consider updating their authentication UI to disallow (or discourage) double-clicking the acceptance button.

**Comment by hempels** [comment]
Link to multi-click issue report on Google Product Forums:
https://groups.google.com/a/googleproductforums.com/d/msgid/apps/92a9fa1a-4bbb-4c93-9e71-2554d7490498%40googleproductforums.com

**Comment by JeanCollas** [comment]
I get the same issue with the Facebook login, with also half of our visitors, so that is really an annoying behavior... We would really appreciate a way to fix this :)

**Comment by hempels** [comment]
As it happens, Google launched a new UI for their Account Chooser page shortly after I attempted (unsuccessfully) to file a bug with someone there who could do something about it. That new UI fixes the double-click problem - after the first click they add an overlay and an activity indicator.

That won't help you any with Facebook if they're having the same issue. I can tell you that in our case, until Google launched their new UI, I replaced the Core Authentication library in our product with a custom build of it that allowed disabling the nonce check. It wasn't particularly hard to implement and I still find it ridiculous that this team refuses to allow developers to use the framework how **_we_** want to rather than how **_they_** want us to. Same old Microsoft mentality in new cross-platform clothes. But I do like these clothes better than the old ones.

https://github.com/aspnet/Security/compare/dev...hempels:rel/1.1.0-preview1
